tests/jax/test_custom_call_compute.py

```diff
@@ -204,8 +204,8 @@ def test_grad_fused_layernorm_fp8_mlp(self, m, n, k,
             b1 = jax.random.normal(subkeys[3], (len(activation_type), n), jnp.bfloat16)
             b2 = jax.random.normal(subkeys[4], (k,), jnp.bfloat16)
         else:
-            b1 = jax.random.normal(subkeys[3], (0,), jnp.bfloat16)
-            b2 = jax.random.normal(subkeys[4], (0,), jnp.bfloat16)
+            b1 = jnp.empty((0,), jnp.bfloat16)
+            b2 = jnp.empty((0,), jnp.bfloat16)
```

Comment:
Does `None` work if there is no_bias? I think `None` may be more used in the real case

Reply:
Hi, as we have a few `reshape` based on `bias.shape` inside the MLP, `None` would break the code.
I would keep those as they are.

Reply:
We can try this but we need to aware that if there is a performance regression with 0 shape bias during the e2e training. I afriad that there are additional all-2-all or grad computations.

Reply:
@zlsh80826 I'm very surprised about this regression.

Why/how does JAX/XLA trigger any extra compute for ops that we have written our custom FWD/BWD implementations? We have full control over what happens in these ops, so I would expect any extra unnecessary compute with an empty bias to be our own bug to fix in TE.

I also don't quite understand how this can trigger extra communication. Our user-facing interfaces for ops/layers with optional biases are written to accept `bias = None` arguments. We internally replace these `None` terms with an *unsharded* `jnp.empty(0)` tensor before calling the underlying custom ops. This should exactly match the `bias_sharding = NamedSharding(mesh, PartitionSpec(None))` we enforce in the primitives' `partition(...)` and `infer_sharding_from_operands(...)` calls, and XLA should not have any reason to trigger extra communication. If it somehow still does trigger extra comms, that suggests either incorrect use of custom partitioning on our end, or incorrect parsing of the primitives in JAX/XLA.

How hard would it be to collect some hard evidence for any extra compute/comm that these empty biases may be triggering in e2e training? It would be great to see an execution profile for a single FWD/BWD pass so we can figure out what (if anything) we need to do in TE custom ops to eliminate them, or otherwise pass the issue over to the JAX/XLA team if it's their bug to fix.

Reply:
Is there any reason why we can't support `biases = [ None, None ]` for `_fused_layernorm_fp8_mlp(...)`?

It should be trivial to replace the check on `use_bias` with checks on `bias_1 is not None` or `bias_2 is not None` during the forward pass, and then save `use_bias_1` and `use_bias_2` flags into the autodiff context so we know what gradients to compute in the backward pass.

Reply:
So, I adjusted the `fused_layernorm_fp8_mlp` so that it can take either `None` or `jnp.empty` as values for biases, providing that the `use_bias = False`.
Could you have a look and let me know if there is an issue?

Reply:
I discovered another issue with zero-shape bias for no-bias cases while resolving the conflict in [PR-796](https://github.com/NVIDIA/TransformerEngine/pull/796). [PR-765](https://github.com/NVIDIA/TransformerEngine/pull/765) began using [zero-shape bias](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/jax/flax/module.py#L1038-L1052) in the MLP module, and I observe that zero-shape bias is created in the Flax model parameters. I believe we should avoid this, as it can confuse users about why biases appear in the model parameters when they did not specify their use. This could also cause the weight converter to fail because the non-TE checkpoint does not include the bias. In [PR-796](https://github.com/NVIDIA/TransformerEngine/pull/796), I had a [WAR](https://github.com/NVIDIA/TransformerEngine/pull/796/commits/5be3f4b1b56756e1db7e435b761fbe733ea32ab6) to remove the reduntant bias in the model parameters. As per @phu0ngng's latest change, the `fused_layernorm_fp8_mlp` can now accept `[None, None]` for no-bias cases. @phu0ngng could you help revise [jnp.empty(0)](https://github.com/NVIDIA/TransformerEngine/pull/796/commits/5be3f4b1b56756e1db7e435b761fbe733ea32ab6#diff-a39a0038aabb6a98b69edbaf40891bf61ec8a77909f175f60128c29b8bbbb1aaR1055-R1056) to `None` after PR-796 merged?

Reply:
I set `bias = None` in the flax modules.